[WIP] add an example notebook explaining hyperparameter optimization using Optuna and showing retrieval on h&m data

radekosmulski opened the pull request:
We have had a couple of discussions recently on exploring using H&M data. Also, some time ago we discussed creating an example around using HPO (it can be a nice way for leveraging raw compute to improve performance, especially in contexts where specialized DL knowledge is not easily accessible or where one doesn't have the time to babysit model training, or even when both of these are available but our business case requires we deliver the best results we can).

The logo is not updated -- if we do end up deciding that this is a notebook we would like to merge, I will generate a proper tracking logo.

rnyak commented on examples/usecases/retrieval-with-hyperparameter-optimization.ipynb:
```diff
@@ -0,0 +1,1405 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #1.    customer_id = ['customer_id'] &gt;&gt; nvt.ops.Categorify() &gt;&gt; nvt.ops.AddMetadata(tags=[Tags.USER_ID])
</pre><p> </p><p> Instead of nvt.ops.AddMetadata(...)  can we use <code>TagAsUserID()</code> and <code>TagAsItemID()</code>  ops?</p><p>you can also get rid of <code>nvt.ops</code> if you do <code>from nvtabular import *</code> at the beginning. </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/786/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

radekosmulski replied:
change made

rnyak commented on examples/usecases/retrieval-with-hyperparameter-optimization.ipynb:
```diff
@@ -0,0 +1,1405 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #2.    train_set_transformed = workflow.fit_transform(train_set)
</pre><p>I'd shorten the names of the processed data sets like <code>train_transformed</code> <code>valid_transformed</code> but not a big deal.  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/786/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

radekosmulski replied:
renamed

rnyak commented on examples/usecases/retrieval-with-hyperparameter-optimization.ipynb:
```diff
@@ -0,0 +1,1405 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #16.        customer_ids_preds = customer_id_mapping.to_pandas().customer_id.iloc[1:].values
</pre><p>why do we need a func for the <code>hit_rate_at_100?</code> is that a required set up by Optuna?  </p><p> </p><p>If not,  cant we use our own metrics class? we already defined <code>recall_at</code> function https://github.com/NVIDIA-Merlin/models/blob/main/merlin/models/tf/metrics/topk.py#L48, so cant we use recall@100? (hit rate and recall should be same).</p><p> </p><p>can we not do like below?</p><p> </p><p><code> metrics = [ mm.RecallAt(100)]</code></p><p><code> model.compile(optimizer, loss=loss, metrics=metrics)</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/786/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

radekosmulski replied:
Yes, you are right. I believe this is needed by Optuna. We need a function that will transform the model into a `top_k_recommender` and that will output a hit count calculated on a per-customer basis (using customer purchases in the validation week and 100 top predictions for a given customer) hence I define this function, to pull together all that functionality.

gabrielspmoreira replied:
<p>Can't you just use model.evaluate() for that retrieval model instead of using <code>model.to_top_k_recommender()</code> ?</p><p>For example</p><p> </p><pre class="ql-syntax" spellcheck="false">eval_metrics = model.evaluate(
        eval_ds,
        batch_size=args.eval_batch_size,
        return_dict=True,
)
</pre><p>Or do you have some specific requirement for your evaluation?</p>

radekosmulski replied:
Thank you very much, Gabriel, for pointing me to this PR 🙏From what I understand, this code is still a month or two away from being merged? Today on the example call I believe we agreed to trying to complete this PR as soon as we can and flagging it for an update when the PR you mentioned goes through as well.

rnyak commented on examples/usecases/retrieval-with-hyperparameter-optimization.ipynb:
```diff
@@ -0,0 +1,1405 @@
+{
```
<p>what does that value correspond to? what's that mean?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/786/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

radekosmulski replied:
<p>That is a great point, Ronay! thank you! will expand on this and add a bit more prose around the interpretation of results</p>

rnyak commented:
@radekosmulski thanks for this notebook. I think this is a useful notebook. I added my comments. I will go ahead one more time after you address my comments. thanks.

rnyak commented on examples/usecases/retrieval-with-hyperparameter-optimization.ipynb:
```diff
@@ -0,0 +1,1405 @@
+{
```
<p>may be remove <code>among</code> from  this <code>how many out of top 100 retrieved candidates were among purchased items</code>  sentence? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/786/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

radekosmulski replied:
good point, removed!

rnyak commented on examples/usecases/retrieval-with-hyperparameter-optimization.ipynb:
```diff
@@ -0,0 +1,1405 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #6.        top_k_rec = model.to_top_k_recommender(item_dataset, 100)
</pre><p>can you pull the latest main branch and  check again if this line is still working? <code>model.to_top_k_recommender(item_dataset, 100)</code>  this  <a href="item_dataset, 100" target="_blank">PR</a> merged this morning, and this might  cause some changes in top_k_recommender. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/786/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

radekosmulski replied:
thank you for highlighting this, just checked, everything still works

radekosmulski commented:
thank you very much for the review, @rnyak! 🙂

I verified that this works with changes from `main`. I also added a bit more prose around interpreting results and a couple of graphs that can be used for diagnostics.

I also implemented the changes you suggested, thank you very much for them!

Unfortunately, this is now failing due to `codespell`. I started a thread on slack, hopefully we can find a resolution.

radekosmulski commented:
The most recent commit adds a tracking logo I created for this nb!

radekosmulski commented:
<p>typo fixed</p>

---

 <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/786/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

bschifferer reviewed (changes requested):
We need a unit test

radekosmulski commented:
<p>added a section that gives the reasoning behind doing HPO, thank you for highlighting this, Benedikt!</p>

---

 <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/786/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

radekosmulski commented:
<p>made the change</p>

---

 <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/786/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

radekosmulski commented:
<p>Added a justification:</p><p> </p><blockquote>This will speed up our experiments (we have less data to train on) and in other experiments that I ran, including all the customers doesn't lead to better results.</blockquote>

---

 <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/786/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

radekosmulski commented:
<p>capture captures output from a cell, otherwise we would have logs from the models being trained and our notebook would become very long</p><p> </p><p>will check with the cI channel re plotly and optuna!</p>

---

 <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/786/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

radekosmulski commented:
<p>Unfortunately, I believe this functionality is missing currently.</p>

---

 <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/786/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

radekosmulski commented:
<p>added a few words of explanation to the NB</p>

---

 <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/786/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

radekosmulski commented:
rerun tests

radekosmulski commented:
rerun tests
